File: 278-FirstBadVersion/278-FirstBadVersion.cpp
// Last updated: 8/1/2025, 2:41:32 PM
bool isBadVersion(int version);

class Solution {
 public:
  int firstBadVersion(int n) {
    int l = 1;
    int r = n;

    while (l < r) {
      const int m = l + (r - l) / 2;
      if (isBadVersion(m))
        r = m;
      else
        l = m + 1;
    }

    return l;
  }
};

File: 278-FirstBadVersion/278-FirstBadVersion.java
// Last updated: 8/1/2025, 2:41:37 PM
public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int start = 1;
        int end = n;
        while (start < end) {

            int mid = start + (end - start) / 2;
            if (isBadVersion(mid)) {
                end = mid;
            } else {
                start = mid + 1;
            }
        }
      
        
        return start;
    }
}
